### ISSUE TITLE
Improve input data size estimation for operations with input ranges

### ISSUE THREAD

**Issue by gritukan** [ytsaurus/ytsaurus#287]
For now, input statistics (row count, data weight, (un)compressed data size) for operations are calculated as a sum of corresponding statistics for input chunks. This is precise for the operations that take whole tables as an input and works pretty well in case if tables and ranges are quite big. However, this approach is very inaccurate for the case of big input chunks and small ranges, since for every range size of the whole chunk will be accounted.

The idea of this proposal is to add a new type of chunk fetcher called range fetcher (feel free to suggest a better naming) that improves estimates for the input data slice statistics by analogy with columnar statistics fetcher. For each chunk that is on the border of some range it fetches block meta from node and calculates the statistics of the blocks that will actually be fetched by a reader. Note, that currently block meta contains row count and uncompressed data size only, so one can add data weight and compressed data size to the chunk meta or just make an approximation `range_data_weight = chunk_data_weight * (range_row_count / chunk_row_count)` that is already used in some other parts of controllers.

**Comment by chegoryu** [comment]
Very nice feature!
I’m now manually selecting the data weight coefficient so that actual input data size for jobs is reasonable.

**Comment by gritukan** [comment]
High-level plan:

1) Introduce an RPC call `GetDataSliceStatistics` on the data node similar to `GetColumnarStatistics`. Its input is a set of chunks, columns and ranges. It returns approximate statistics of what will be read (total size of blocks in the projection of the chunk to provided columns and keys).
2) Implement TDataSliceStatisticsFetcher similar to columnar statistics fetcher that fetches statistics from nodes for given set of chunks.
3) Use fetcher in controller to make data size and row count estimations more precise. This feature should be opt-in (at least for the first time)

**Comment by zlobober** [comment]
Rpc call should probably be named GetChunkSliceStatistics instead of GetDataSliceStatistics, because there is no notion of data slice in data node (data slice if a logical union of a number of overlapping chunk slices of different chunks). Otherwise, I'm ok with the plan

**Comment by gritukan** [comment]
Yeah, you are right. It's chunk slices in relevant places of code.

**Comment by gritukan** [comment]
Also, there is a room for possible optimization: if chunk is completely inside some range we call do a fallback to columnar statistics fetching which can be done from masters. However, I think it is not required for the first version.
